File: backend/openrouter.py
"""OpenRouter API client (stdlib-only HTTP)."""

import json
import urllib.request
import urllib.error
from typing import Any, Dict, List, Optional

from .config import OPENROUTER_KEY, API_URL


def _content_from_parts(content: Any) -> Optional[str]:
    if isinstance(content, str) and content:
        return content
    if isinstance(content, dict):
        return json.dumps(content)
    if isinstance(content, list):
        parts = []
        for item in content:
            if (
                isinstance(item, dict)
                and item.get("type") in {"text", "output_text"}
            ):
                text = item.get("text")
                if isinstance(text, str):
                    parts.append(text)
        if parts:
            return "".join(parts)
    return None


def _extract_parsed_json(response: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    if not isinstance(response, dict):
        return None

    if isinstance(response.get("parsed"), dict):
        return response.get("parsed")

    choices = response.get("choices", [])
    if not choices or not isinstance(choices[0], dict):
        return None
    message = choices[0].get("message", {})
    if not isinstance(message, dict):
        return None

    parsed = message.get("parsed")
    if isinstance(parsed, dict):
        return parsed

    content = message.get("content")
    if isinstance(content, dict):
        return content
    return None


def _extract_content(response: Dict[str, Any]) -> Optional[str]:
    if not isinstance(response, dict):
        return None
    content = _content_from_parts(response.get("content"))
    if content:
        return content
    choices = response.get("choices", [])
    if not choices:
        return None
    first = choices[0] if isinstance(choices[0], dict) else {}
    message = first.get("message", {}) if isinstance(first.get("message"), dict) else {}
    content = _content_from_parts(message.get("content"))
    if content:
        return content
    return first.get("text") if isinstance(first.get("text"), str) else None


def _strip_code_fences(text: str) -> str:
    stripped = text.strip()
    if stripped.startswith("```"):
        stripped = stripped.split("\n", 1)[-1]
    if stripped.endswith("```"):
        stripped = stripped.rsplit("\n", 1)[0]
    return stripped.strip()


def _parse_json_content(content: str) -> Optional[Dict[str, Any]]:
    if not content:
        return None
    text = _strip_code_fences(content)
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        start = text.find("{")
        end = text.rfind("}")
        if start == -1 or end == -1 or end <= start:
            return None
        try:
            return json.loads(text[start : end + 1])
        except json.JSONDecodeError:
            return None


def call_openrouter(
    model: str,
    messages: List[Dict[str, str]],
    *,
    max_tokens: Optional[int] = None,
    reasoning: Optional[Dict[str, Any]] = None,
    response_format: Optional[Dict[str, Any]] = None,
    provider: Optional[Dict[str, Any]] = None,
    api_url: Optional[str] = None,
    timeout: float = 120.0,
    headers: Optional[Dict[str, str]] = None,
) -> Optional[Dict[str, Any]]:
    """Call OpenRouter with chat messages and return raw JSON response."""
    if not OPENROUTER_KEY:
        print("OPENROUTER_API_KEY is missing.")
        return None

    target_url = api_url or API_URL
    if not target_url:
        print("OPENROUTER_API_URL is missing.")
        return None

    request_headers = {
        "Authorization": f"Bearer {OPENROUTER_KEY}",
        "Content-Type": "application/json",
        **(headers or {}),
    }

    payload: Dict[str, Any] = {"model": model, "messages": messages}
    if max_tokens is not None:
        payload["max_tokens"] = max_tokens
    if reasoning is not None:
        payload["reasoning"] = reasoning
    if response_format is not None:
        payload["response_format"] = response_format
    if provider is not None:
        payload["provider"] = provider

    try:
        data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            target_url,
            data=data,
            headers=request_headers,
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = ""
        try:
            body = e.read().decode("utf-8", errors="replace")
        except Exception:
            body = ""
        print(f"Error querying OpenRouter: HTTP {e.code} {e.reason}")
        if body:
            print(f"OpenRouter error body: {body}")
        return {
            "error": "HTTPError",
            "status": e.code,
            "reason": str(e.reason),
            "body": body,
        }
    except urllib.error.URLError as e:
        print(f"Error querying OpenRouter: {e}")
        return {
            "error": "URLError",
            "reason": str(getattr(e, "reason", e)),
        }
    except Exception as e:
        print(f"Unexpected error querying OpenRouter: {e}")
        return {"error": "UnexpectedError", "message": str(e)}


def parse_openrouter_json(response: Dict[str, Any]) -> tuple[Optional[Dict[str, Any]], str]:
    """Extract JSON content from an OpenRouter response."""
    parsed_json = _extract_parsed_json(response)
    if isinstance(parsed_json, dict):
        return parsed_json, json.dumps(parsed_json)

    content = _extract_content(response) or ""
    parsed = _parse_json_content(content)
    return parsed, content

File: lib/backend/openrouter.ts
import { OPENROUTER_CONFIG } from "@/lib/backend/config";
import type { UnknownRecord } from "@/lib/backend/types";

type OpenRouterMessage = {
  role: string;
  content: string;
};

type CallOpenRouterOptions = {
  max_tokens?: number;
  reasoning?: UnknownRecord;
  response_format?: UnknownRecord;
  provider?: UnknownRecord;
  api_url?: string;
  timeout?: number;
  headers?: Record<string, string>;
};

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === "object" && value !== null;
}

function contentFromParts(content: unknown): string | undefined {
  if (typeof content === "string" && content) {
    return content;
  }
  if (isRecord(content)) {
    return JSON.stringify(content);
  }
  if (Array.isArray(content)) {
    const parts: string[] = [];
    for (const item of content) {
      if (!isRecord(item)) {
        continue;
      }
      const itemType = item.type;
      const text = item.text;
      if (
        (itemType === "text" || itemType === "output_text") &&
        typeof text === "string"
      ) {
        parts.push(text);
      }
    }
    if (parts.length > 0) {
      return parts.join("");
    }
  }
  return undefined;
}

function extractParsedJson(response: unknown): UnknownRecord | undefined {
  if (!isRecord(response)) {
    return undefined;
  }
  if (isRecord(response.parsed)) {
    return response.parsed;
  }

  const choices = response.choices;
  if (!Array.isArray(choices) || choices.length === 0 || !isRecord(choices[0])) {
    return undefined;
  }
  const message = choices[0].message;
  if (!isRecord(message)) {
    return undefined;
  }

  if (isRecord(message.parsed)) {
    return message.parsed;
  }
  if (isRecord(message.content)) {
    return message.content;
  }
  return undefined;
}

function extractContent(response: unknown): string {
  if (!isRecord(response)) {
    return "";
  }

  const topLevel = contentFromParts(response.content);
  if (topLevel) {
    return topLevel;
  }

  const choices = response.choices;
  if (!Array.isArray(choices) || choices.length === 0 || !isRecord(choices[0])) {
    return "";
  }
  const first = choices[0];
  const message = isRecord(first.message) ? first.message : {};
  const fromMessage = contentFromParts(message.content);
  if (fromMessage) {
    return fromMessage;
  }
  return typeof first.text === "string" ? first.text : "";
}

function stripCodeFences(text: string): string {
  let stripped = text.trim();
  if (stripped.startsWith("```")) {
    stripped = stripped.split("\n", 2)[1] ?? "";
  }
  if (stripped.endsWith("```")) {
    const lines = stripped.split("\n");
    lines.pop();
    stripped = lines.join("\n");
  }
  return stripped.trim();
}

function parseJsonContent(content: string): UnknownRecord | undefined {
  if (!content) {
    return undefined;
  }

  const text = stripCodeFences(content);
  try {
    const parsed = JSON.parse(text);
    return isRecord(parsed) ? parsed : undefined;
  } catch {}

  const start = text.indexOf("{");
  const end = text.lastIndexOf("}");
  if (start < 0 || end < 0 || end <= start) {
    return undefined;
  }
  try {
    const parsed = JSON.parse(text.slice(start, end + 1));
    return isRecord(parsed) ? parsed : undefined;
  } catch {
    return undefined;
  }
}

export async function callOpenRouter(
  model: string,
  messages: OpenRouterMessage[],
  options: CallOpenRouterOptions = {},
): Promise<UnknownRecord | null> {
  const apiKey = OPENROUTER_CONFIG.OPENROUTER_API_KEY;
  if (!apiKey) {
    return null;
  }
  const targetUrl = options.api_url ?? OPENROUTER_CONFIG.OPENROUTER_API_URL;
  if (!targetUrl) {
    return null;
  }

  const payload: UnknownRecord = { model, messages };
  if (options.max_tokens !== undefined) {
    payload.max_tokens = options.max_tokens;
  }
  if (options.reasoning !== undefined) {
    payload.reasoning = options.reasoning;
  }
  if (options.response_format !== undefined) {
    payload.response_format = options.response_format;
  }
  if (options.provider !== undefined) {
    payload.provider = options.provider;
  }

  const controller = new AbortController();
  const timeoutMs = (options.timeout ?? 120) * 1000;
  const timeout = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(targetUrl, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
        ...(options.headers ?? {}),
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
      cache: "no-store",
    });

    const text = await response.text();
    let parsed: unknown = {};
    if (text) {
      try {
        parsed = JSON.parse(text) as unknown;
      } catch {
        parsed = { content: text };
      }
    }
    if (!response.ok) {
      return {
        error: "HTTPError",
        status: response.status,
        reason: response.statusText,
        body: text,
      };
    }
    return isRecord(parsed) ? parsed : { raw: parsed, content: text };
  } catch (error) {
    return {
      error: "UnexpectedError",
      message: error instanceof Error ? error.message : String(error),
    };
  } finally {
    clearTimeout(timeout);
  }
}

export function parseOpenRouterJson(response: unknown): {
  parsed: UnknownRecord | undefined;
  content: string;
} {
  const parsedJson = extractParsedJson(response);
  if (parsedJson) {
    return { parsed: parsedJson, content: JSON.stringify(parsedJson) };
  }

  const content = extractContent(response);
  const parsed = parseJsonContent(content);
  return { parsed, content };
}
